cmd/end2endtest/account.go

```diff
@@ -185,35 +215,26 @@ func testSendTokens(api *apiclient.HTTPclient, aliceKeys, bobKeys *ethereum.Sign
 	log.Debugf("mined, tx refs are %+v and %+v", txrefa, txrefb)
 
 	// now check the resulting state
-	bobAccAfter, err := bob.Account("")
-	if err != nil {
-		return err
-	}
-	if bobAccAfter == nil {
-		return state.ErrAccountNotExist
-	}
-	log.Infow("bob", "account", bobKeys.Address(), "nonce", bobAcc.Nonce, "balance", bobAcc.Balance)
-	if bobAcc.Balance-23-uint64(txCost)+19 != bobAccAfter.Balance {
-		log.Fatalf("expected %s to have balance %d got %d", bobKeys.Address(), 300, bobAccAfter.Balance)
-	}
+	checkAccountNonceAndBalance(alice, aliceAcc.Nonce+1,
+		(aliceAcc.Balance - amountAtoB - uint64(txCost) + amountBtoA))
+	checkAccountNonceAndBalance(bob, bobAcc.Nonce+1,
+		(bobAcc.Balance - amountBtoA - uint64(txCost) + amountAtoB))
+	return nil
+}
 
-	aliceAccAfter, err := alice.Account("")
+func checkAccountNonceAndBalance(api *apiclient.HTTPclient, expNonce uint32, expBalance uint64) {
+	acc, err := api.Account("")
 	if err != nil {
-		return err
-	}
-	if aliceAccAfter == nil {
-		return state.ErrAccountNotExist
+		log.Fatal(err)
 	}
-	log.Infow("alice", "account", aliceKeys.Address(), "nonce", aliceAcc.Nonce, "balance", aliceAcc.Balance)
-	if aliceAcc.Balance-19-uint64(txCost)+23 != aliceAccAfter.Balance {
-		log.Fatalf("expected %s to have balance %d got %d",
-			aliceKeys.Address(), aliceAcc.Balance-(100+uint64(txCost)), aliceAccAfter.Balance)
+	log.Infow("current state", "account", acc.Address.String(), "nonce", acc.Nonce, "balance", acc.Balance)
+
+	if expBalance != acc.Balance {
+		log.Fatalf("expected %s to have balance %d got %d", acc.Address.String(), expBalance, acc.Balance)
 	}
-	if aliceAcc.Nonce+1 != aliceAccAfter.Nonce {
-		log.Fatalf("expected %s to have nonce %d got %d",
-			aliceKeys.Address(), aliceAcc.Nonce+1, aliceAccAfter.Nonce)
+	if expNonce != acc.Nonce {
+		log.Fatalf("expected %s to have nonce %d got %d", acc.Address.String(), expNonce, acc.Nonce)
```

Comment:
I don't think none of these functions should fatal. They should return error and be correctly propagated until the initial caller of the test.

Reply:
in general i'd agree, but in my experience with end2endtest it helps A LOT to call Fatal on the relevant line instead of returning the error and Fatal on the caller, because this obscures the source of the error.
being an integration test, i think we can make these kind of compromises for the sake of readability or debugging

on another hand, i just realized that since zerolog implementation, Fatal no longer spits a stack trace, which was tremendously useful precisely to handle this situation i just wrote.

Reply:
i'll keep all the Fatals as-is for now, and consider this change for a second round of refactoring, to not make this PR eternal